File: CHUOND_4/LearnIntent/app/src/main/java/com/example/i/MainActivity.java
package com.example.i;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.i.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {
    public static final String EXTRA_FULL_NAME = "extra_full_name";
    public static final String EXTRA_AGE = "extra_age";

    private ActivityMainBinding mBinding;
    @SuppressLint("UnsafeImplicitIntentLaunch")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
        ViewCompat.setOnApplyWindowInsetsListener(mBinding.main, (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        mBinding.btnSubmit.setOnClickListener(v -> {
           Intent intent = new Intent(this, SampleActivity.class);
           Bundle bundle = new Bundle();
           String fullName = String.valueOf(mBinding.txtFullName.getText());
           String ageStr = String.valueOf(mBinding.txtAge.getText());
           int age = 0;
           if (!ageStr.isEmpty()){
               age = Integer.parseInt(ageStr);
           }
           bundle.putInt(EXTRA_AGE, age);
           bundle.putString(EXTRA_FULL_NAME, fullName);
           intent.putExtras(bundle);
           startActivity(intent);
        });

//        mBinding.btnLaunchActivity.setOnClickListener(v -> {
//            Intent intent = new Intent();
//            intent.setAction(Intent.ACTION_VIEW);
//            intent.addCategory(Intent.CATEGORY_DEFAULT);
//            startActivity(intent);
//        });
    }
}

File: CHUONG_7/Ex1_RetrofitHelper/app/src/main/java/com/example/myapplication/MainActivity.kt
package com.example.myapplication

import android.os.Bundle
import android.view.View
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.myapplication.databinding.ActivityMainBinding
import com.example.myapplication.ui.UserAdapter
import com.example.myapplication.ui.UserViewModel

class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private val adapter = UserAdapter()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        binding.recycler.layoutManager = LinearLayoutManager(this)
        binding.recycler.adapter = adapter
        binding.recycler.addItemDecoration(
            DividerItemDecoration(this, DividerItemDecoration.VERTICAL)
        )
        val vm = ViewModelProvider(this)[UserViewModel::class.java]

        binding.progress.visibility = View.VISIBLE
        vm.loadUsers()

        vm.users.observe(this) { list ->
            binding.progress.visibility = View.GONE
            binding.errorGroup.visibility = View.GONE
            adapter.submit(list)
        }
        vm.error.observe(this) { msg ->
            if (msg != null) {
                binding.progress.visibility = View.GONE
                binding.errorGroup.visibility = View.VISIBLE
                binding.textError.text = msg
            }
        }

        binding.btnRetry.setOnClickListener {
            binding.progress.visibility = View.VISIBLE
            binding.errorGroup.visibility = View.GONE
            vm.loadUsers()
        }
    }
}